target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,506 @@
+
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) & ~(1 << 12), RTL838X_PLL_GLB_CTRL);
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) & ~1, RTL838X_PLL_GLB_CTRL);
+
+	sw_w32(reg->misc, RTL838X_PLL_CPU_MISC_CTRL);
+	sw_w32(reg->ctrl0, RTL838X_PLL_CPU_CTRL0);
+	sw_w32(reg->ctrl1, RTL838X_PLL_CPU_CTRL1);
+
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) | 1, RTL838X_PLL_GLB_CTRL);
+	while(!(sw_r32(RTL838X_PLL_GLB_CTRL) & (1 << 8)));
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) | (1 << 12), RTL838X_PLL_GLB_CTRL);
+
+	spin_unlock_irqrestore(&rtl_ccu->lock, flags);
+
+	return 0;
+}
+
+static int rtl839x_set_rate(int clk_idx, const struct rtl_reg_set *reg)
+{
+	unsigned long flags;
+	void *fnaddr;
+	unsigned int fnsize;
+	void (*fnsram)(unsigned int clk_idx, unsigned int ctrl0, unsigned int ctrl1);
+
+	switch (clk_idx) {
+	case CLK_CPU:
+	case CLK_MEM:
+	case CLK_LXB:
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	spin_lock_irqsave(&rtl_ccu->lock, flags);
+/*
+ * That might be creepy but we cannot change clocks on RTL839X SOC from normal
+ * memory. Relocate the setter function to the start of SRAM and execute it from
+ * there so the system will not freeze.
+ */
+	fnsram = (void *)RTL839X_SRAM_BASE;
+	fnaddr = &rtl839x_set_rate_sram;
+	fnsize = &rtl839x_set_rate_sram_end - &rtl839x_set_rate_sram;
+
+	memcpy(fnsram, fnaddr, fnsize);
+	flush_icache_range((unsigned long)fnsram, (unsigned long)(fnsram + fnsize));
+	fnsram(clk_idx, reg->ctrl0, reg->ctrl1);
+
+	spin_unlock_irqrestore(&rtl_ccu->lock, flags);
+
+	return 0;
+}
+
+static int rtl_set_rate(struct clk_hw *hw, unsigned long rate, unsigned long parent_rate)
+{
+	struct rtl_clk *clk = rtl_hw_to_clk(hw);
+	int tab_idx;
+	const struct rtl_rtab_set *rtab;
+
+	if (parent_rate != OSC_RATE)
+		return -EINVAL;
+
+	rtab = &rtl_rtab_set[rtl_ccu->soc][clk->idx];
+
+	tab_idx = (rate - rtl_round_set[rtl_ccu->soc][clk->idx].min)
+		/ rtl_round_set[rtl_ccu->soc][clk->idx].step; 
+
+	if (tab_idx < 0 || tab_idx >= rtab->count)
+		return -EINVAL;
+
+	switch (rtl_ccu->soc) {
+	case SOC_RTL838X:
+		return rtl838x_set_rate(clk->idx, &rtab->rset[tab_idx]);
+	case SOC_RTL839X:
+		return rtl839x_set_rate(clk->idx, &rtab->rset[tab_idx]);
+	default:
+		return -ENXIO;
+	}
+}
+
+static long rtl_round_rate(struct clk_hw *hw, unsigned long rate, unsigned long *parent_rate)
+{
+	struct rtl_clk *clk = rtl_hw_to_clk(hw);
+	const struct rtl_round_set *round = &rtl_round_set[rtl_ccu->soc][clk->idx];
+	unsigned long rrate;
+
+	rrate = max(round->min, rate);
+	rrate = min(round->max, rrate);
+	return round->min + (((rrate - round->min + round->step - 1) / round->step) * round->step);
+}
+
+/*
+ * Initialization functions to register the CCU (core clock unit)
+ */
+
+static const struct clk_ops rtl_clk_ops = {
+	.round_rate = rtl_round_rate,
+	.recalc_rate = rtl_recalc_rate,
+	.set_rate = rtl_set_rate,
+};
+
+static int rtl_ccu_create(struct device_node *np)
+{
+	if (soc_info.family != RTL8380_FAMILY_ID && soc_info.family != RTL8390_FAMILY_ID)
+		return -ENXIO;
```

Comment:
If you don't use `soc_info`, this can be dropped.

Reply:
fixed in upcoming version